File: ios/MoWalk/Models/TrendsViewModel.swift
import Observation
import SwiftUI

enum TrendsPeriod: String, CaseIterable {
    case week = "This Week"
    case month = "This Month"
}

@MainActor
@Observable
final class TrendsViewModel {
    var dataPoints: [StepEntry] = []
    var period: TrendsPeriod = .week
    var selectedDay: StepEntry?
    var isLoading: Bool = false

    private let persistence: PersistenceService

    init(persistence: PersistenceService) {
        self.persistence = persistence
        loadData()
    }

    func setPeriod(_ newPeriod: TrendsPeriod) {
        period = newPeriod
        loadData()
    }

    func loadData() {
        isLoading = true
        defer { isLoading = false }

        do {
            switch period {
            case .week:
                dataPoints = try persistence.getWeeklyEntries()
            case .month:
                let calendar = Calendar.current
                let now = Date()
                let year = calendar.component(.year, from: now)
                let month = calendar.component(.month, from: now)
                dataPoints = try persistence.getMonthlyEntries(year: year, month: month)
            }
        } catch {
            dataPoints = []
        }
    }

    func selectDay(_ day: StepEntry?) {
        selectedDay = day
    }
}

File: android/app/src/main/java/com/mowalk/app/ui/trends/TrendsViewModel.kt
package com.mowalk.app.ui.trends

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.mowalk.app.data.local.DailyStepEntity
import com.mowalk.app.data.repository.StepRepository
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch

data class TrendsState(
    val dataPoints: List<DailyStepEntity> = emptyList(),
    val period: TrendsPeriod = TrendsPeriod.WEEK,
    val isLoading: Boolean = false,
    val selectedDay: DailyStepEntity? = null
)

enum class TrendsPeriod {
    WEEK, MONTH
}

class TrendsViewModel(
    private val repository: StepRepository
) : ViewModel() {

    private val _state = MutableStateFlow(TrendsState())
    val state: StateFlow<TrendsState> = _state.asStateFlow()

    init {
        loadWeeklyData()
    }

    fun setPeriod(period: TrendsPeriod) {
        _state.value = _state.value.copy(period = period)
        when (period) {
            TrendsPeriod.WEEK -> loadWeeklyData()
            TrendsPeriod.MONTH -> loadMonthlyData()
        }
    }

    fun onSelectDay(day: DailyStepEntity) {
        _state.value = _state.value.copy(selectedDay = day)
    }

    fun onDismissDetail() {
        _state.value = _state.value.copy(selectedDay = null)
    }

    private fun loadWeeklyData() {
        viewModelScope.launch {
            try {
                repository.getWeeklyData().collect { data ->
                    _state.value = _state.value.copy(dataPoints = data)
                }
            } catch (e: Exception) {
                _state.value = _state.value.copy(isLoading = false)
            }
        }
    }

    private fun loadMonthlyData() {
        val now = java.time.LocalDate.now()
        viewModelScope.launch {
            try {
                repository.getMonthlyData(now.year, now.monthValue).collect { data ->
                    _state.value = _state.value.copy(dataPoints = data)
                }
            } catch (e: Exception) {
                _state.value = _state.value.copy(isLoading = false)
            }
        }
    }
}
